### PR TITLE
Improve acronym function names

### PR DESCRIPTION
**Pull request by jplewa**
## Task

Resolves: #221 

## Description

`IAMBinding` -> `iamBinding`
`IamAuditConfig` -> `iamAuditConfig`
`OrganizationPolicy` -> `organizationPolicy`

This one is a bit awkward, but it's not our fault they called it `VMware` instead of `Vmware`:
`VMwareCluster` -> `vMwareCluster`

### CODE REVIEW

**Review by myhau — APPROVED** [review]

**Comment by myhau** [comment]
For the record, here is the list with "tricky" type/resource names for providers supported by `pulumi-kotlin` (kudos to @jplewa for generating it):

<details><summary>Open the list</summary>
<p>

```
AAAARecord
AADDataConnector
AATPDataConnector
ACIService
ACL
ADCCatalog
ADLSGen1FileDataSet
ADLSGen1FolderDataSet
ADLSGen2FileDataSet
ADLSGen2FileDataSetMapping
ADLSGen2FileSystemDataSet
ADLSGen2FileSystemDataSetMapping
ADLSGen2FolderDataSet
ADLSGen2FolderDataSetMapping
AFDCustomDomain
AFDEndpoint
AFDOrigin
AFDOriginGroup
AKSService
APIService
APIServiceList
APIServicePatch
ARecord
ASCDataConnector
AScript
CACertificate
CNameRecord
CRL
CSIDriver
CSIDriverList
CSIDriverPatch
CSINode
CSINodeList
CSINodePatch
CSIStorageCapacity
CSIStorageCapacityList
CSIStorageCapacityPatch
DBAuditInstance
DBCluster
DBClusterLakeVersion
DBClusterParameterGroup
DBInstance
DBParameterGroup
DBProxy
DBProxyEndpoint
DBProxyTargetGroup
DBSubnetGroup
DHCPOptions
DNSSEC
DRTAccess
EC2Fleet
EIP
FHIRDatastore
GCPolicy
HAVip
HAVipAttachment
HBaseCluster
IAMAuditConfig
IAMBinding
IAMCustomRole
IAMMember
IAMPolicy
IPAM
IPAMAllocation
IPAMPool
IPAMPoolCidr
IPAMResourceDiscovery
IPAMResourceDiscoveryAssociation
IPAMScope
IPAddress
IPAddressList
IPAddressPatch
IPGroup
IPGroupCIDR
IPSet
IPv6FirewallRule
IVPCEConfiguration
MCASDataConnector
MDATPDataConnector
MECRole
MHSMPrivateEndpointConnection
MLTransform
MSIXPackage
OIDCProvider
PTRRecord
SAMLProvider
SAPApplicationServerInstance
SAPCentralInstance
SAPDatabaseInstance
SAPVirtualInstance
SRVRecord
SSLCertificate
SSLPolicy
TIDataConnector
URLMap
VCenter
VMWareReplicationPolicy
VMwareCluster
VMwareCollector
VMwareNodePool
VNetPeering
VPC
VPCConnection
VPCDHCPOptionsAssociation
VPCEConfiguration
VPCEndpoint
VPCEndpointService
VPCEndpointServicePermissions
VPCPeeringConnection
VPNConnection
VPNConnectionRoute
VPNGateway
VPNTunnel
WCFRelay
WCFRelayAuthorizationRule
```

</p>
</details> 

I'm wondering how we handle these:
- `IPv6FirewallRule`
- `IPGroupCIDR`
- `EC2Fleet`

**Comment by jplewa** [comment]
> I'm wondering how we handle these:
>   - `IPv6FirewallRule`
>   - `IPGroupCIDR`
>   - `EC2Fleet`

`IPv6FirewallRule` -> `iPv6FirewallRule`
`IPGroupCIDR` -> `ipGroupCIDR`
`EC2Fleet` -> `eC2Fleet`

The first two seem fine to me. `iPv6FirewallRule` is a bit awkward, but I'd say it's correct looking at the actual class name. Tbh, the actual class name is weird to me, since I'd rather call the class `Ipv6FirewallRule` (see: https://stackoverflow.com/questions/34529999/camel-case-name-for-class-with-ipv4). The only one I'm wondering about is `EC2Fleet`. Maybe the rule should be "make upper case letters lowercase until you encounter the final upper-case letter **or a number**"? So that we end up with `ec2Fleet`?

File: src/test/kotlin/org/virtuslab/pulumikotlin/codegen/step2intermediate/PulumiNameTest.kt [hunk 1]
```diff
@@ -555,4 +555,49 @@ internal class PulumiNameTest {
...
+        // when
+        val pulumiName = PulumiName.from(token, namingConfiguration)
+        val namingFlags = NamingFlags(Root, Resource, Input, Java)
+        val resourceFunctionName = pulumiName.toResourceFunctionName(namingFlags)
+
+        // then
+        assertEquals("iamBinding", resourceFunctionName)
+    }
+
+    @Test
```
Comment (myhau) [inline src/test/kotlin/org/virtuslab/pulumikotlin/codegen/step2intermediate/PulumiNameTest.kt:589]: Could you add a test case for full upper-caps names like `CRL`?

File: src/test/kotlin/org/virtuslab/pulumikotlin/codegen/utils/StringUtilsKtTest.kt [hunk 2]
```diff
@@ -27,6 +27,11 @@ internal class StringUtilsKtTest {
         assertEquals("word", "Word".decapitalize())
     }
 
+    @Test
+    fun `should turn the expected number of character lowercase`() {
+        assertEquals("sslCertificate", "SSLCertificate".lowercaseBefore(3))
+    }
```
Comment (myhau) [inline src/test/kotlin/org/virtuslab/pulumikotlin/codegen/utils/StringUtilsKtTest.kt:33]: Covering edge cases like `lowercaseBefore(string.length)`, `lowercaseBefore(1)` and `lowercaseBefore(0)` would be nice (and cheap).

File: src/test/kotlin/org/virtuslab/pulumikotlin/codegen/step2intermediate/PulumiNameTest.kt [hunk 3]
```diff
@@ -555,4 +555,64 @@ internal class PulumiNameTest {
...
+        val token = "provider:module:EC2Fleet"
+        val namingConfiguration = PulumiNamingConfiguration.create(providerName = "provider")
+
+        // when
+        val pulumiName = PulumiName.from(token, namingConfiguration)
+        val namingFlags = NamingFlags(Root, Resource, Input, Java)
+        val resourceFunctionName = pulumiName.toResourceFunctionName(namingFlags)
+
+        // then
+        assertEquals("ec2Fleet", resourceFunctionName)
```
Comment (myhau) [inline src/test/kotlin/org/virtuslab/pulumikotlin/codegen/step2intermediate/PulumiNameTest.kt:616]: Adding one more test for `VMwareCluster` might be nice, so that we are aware of this edge case.

**Review by myhau — APPROVED** [review]
LGTM. Left a couple of minor comments, but feel free to merge now.

File: src/main/kotlin/org/virtuslab/pulumikotlin/codegen/step2intermediate/PulumiName.kt [hunk 4]
```diff
@@ -202,6 +203,18 @@ data class PulumiName(
...
+    fun toResourceFunctionName(namingFlags: NamingFlags): String {
+        val resourceName = toResourceName(namingFlags)
+        val numberOfCaps = resourceName.takeWhile { it.isUpperCase() || it.isDigit() }.count()
+        return if (numberOfCaps == resourceName.length) {
+            resourceName.lowercase()
+        } else if (numberOfCaps > 1) {
+            resourceName.lowercaseBefore(numberOfCaps - 1)
+        } else {
+            resourceName.decapitalize()
+        }
```
[the remark this answers is not in the document]
Comment (jplewa) [reply src/main/kotlin/org/virtuslab/pulumikotlin/codegen/step2intermediate/PulumiName.kt:215]: For `CRL`, `numberOfCaps=3` and if we do `"CRL".lowercaseBefore(2)`, that returns the string `crL`. This is an edge case where all characters are upper-case and we don't want to skip this final upper-case character.

For `OrganizationPolicy`, `numberOfCaps=1`, so we end up doing `"OrganizationPolicy".lowercaseBefore(0), which doesn't turn any of the characters to lowercase. This is the other edge case where there's only one upper-case character and we still want to turn it to lower case.

What we could do is this:
```kotlin
        return if (numberOfCaps == resourceName.length || numberOfCaps == 1) {
            resourceName.lowercaseBefore(numberOfCaps)
        } else {
            resourceName.lowercaseBefore(numberOfCaps - 1)
        }
```
but I'm not sure if that's easier to read.

**Comment by jplewa** [comment]
> LGTM. Left a couple of minor comments, but feel free to merge now.

Can't merge, because this is branched off of https://github.com/VirtuslabRnD/pulumi-kotlin/pull/241 ;)

File: src/main/kotlin/org/virtuslab/pulumikotlin/codegen/step2intermediate/PulumiName.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (myhau) [reply src/main/kotlin/org/virtuslab/pulumikotlin/codegen/step2intermediate/PulumiName.kt:216]: ok, leave it as is then :)
